### PR TITLE
soc/arm/st_stm32/stm32l4:Add power management and shutdown mode

### PR DESCRIPTION
**Pull request by khalidjmi1200**
Add support for power management and the shutdown mode for stm32l4.
This now allows boards that have any kind of low power mode to send
it to shutdown mode and wake it up from wakeup source.
Currently wake up source it set as PIN1.

Tested on a STM32L462VE for shutdown and wakeup.

Signed-off-by: Khalid Mustafa <khalid.mustafa@greyp.com>

### CODE REVIEW

File: soc/arm/st_stm32/stm32l4/power.c [hunk 1]
```diff
@@ -26,18 +26,38 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
...
+}
 
-	switch (info.substate_id) {
+static void pm_state_set_suspend_to_idle(uint8_t substate_id)
+{
+	switch (substate_id) {
 	case 1: /* this corresponds to the STOP0 mode: */
 		/* ensure the proper wake-up system clock */
-		LL_RCC_SetClkAfterWakeFromStop(RCC_STOP_WAKEUPCLOCK_SELECTED);
+		LL_RCC_SetClkAfterWakeFromStop(
```
Comment (tagunil) [inline soc/arm/st_stm32/stm32l4/power.c:59]: Why are these changes needed? The current line length limit is about 100 chars.

File: soc/arm/st_stm32/stm32l4/power.c [hunk 2]
```diff
@@ -26,15 +26,34 @@ LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
...
-void pm_power_state_set(struct pm_state_info info)
+static void pm_state_set_soft_off(uint8_t substate_id)
 {
-	if (info.state != PM_STATE_SUSPEND_TO_IDLE) {
-		LOG_DBG("Unsupported power state %u", info.state);
-		return;
+	switch (substate_id) {
+	case 1: /* this corresponds to the Shutdown mode: */
+		/* Disable Pin 1 which is used as wakeup source */
+		LL_PWR_DisableWakeUpPin(LL_PWR_WAKEUP_PIN1);
```
[the remark this answers is not in the document]
Comment (erwango) [reply soc/arm/st_stm32/stm32l4/power.c:34]: This needs to be configurable. Can you check how to configure this from device tree ?

[the remark this answers is not in the document]
Comment (khalidjmi1200) [reply soc/arm/st_stm32/stm32l4/power.c:34]: Can you please let me know what exactly you wanted to be configured or any reference point?
I tried searching for the resources related to this pin configuration but haven't  found anything.

File: soc/arm/st_stm32/stm32l4/power.c [hunk 3]
```diff
@@ -79,11 +113,11 @@ void pm_power_state_exit_post_ops(struct pm_state_info info)
 		LOG_DBG("Unsupported power substate-id %u", info.state);
```
Comment (erwango) [inline soc/arm/st_stm32/stm32l4/power.c:113]: Don't we need to update this function now?

File: soc/arm/st_stm32/stm32l4/power.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (erwango) [reply soc/arm/st_stm32/stm32l4/power.c:34]: ok, so I haven't found anything quick and easy to implement.
Mostly it would require preliminary changes in device tree that would take some time to think, implement and finally get merged.
This being said, I'd like to avoid merging code that only fit a particular implementation.

I'm thinking to an alternate proposal that would each one to test this and adapt easily on his use case.
Would you mind:
-Removing the code part that handle pin in this file
-Add a dedicated power_off sample under `samples/boards/stm32/power_mgmt` that would demonstrate use of one convenient PIN as a wake up. Ideally the sample would work on a in tree board.

**Review by erwango — CHANGES_REQUESTED** [review]

**Comment by erwango** [comment]
@khalidjmi1200, About the wake up pin configuration, please note that adding the sample demonstrating use of wake up pin is not blocking. The only blocking point is that code should not be added in power.c (as long as we don't have proper dts support for it).
The sample is just a nice bonus.
